docs/src/proposals/interest-bearing-tokens.md

```diff
@@ -0,0 +1,284 @@
+interacting with. See the Associated Token Account section for how to use this.
+
+There are also new read-only instructions, which write data to a provided account
+buffer:
+
+* `GetHoldingBalance`: balance of the holding
+* `GetHoldingOwner`: owner's public key
+* `GetMintSupply`: mint supply
+* `GetMintAuthority`: mint authority's public key
+* `GetMintDecimals`: mint decimals
+
+See the Runtime section for more information on how these are used.
+
+Programs and wallets that wish to support i-tokens must update the instructions
+to use all `Checked` variants and use the new read-only instructions to fetch
+information.
+
+The SPL token library provides wrappers compatible with any conforming program.
+For the base SPL token program, the wrapper simply deserializes the account data
+and returns the relevant data. For other programs, it calls the appropriate
+read-only instruction using an ephemeral account input.
+
+The structure for new holding and mint accounts must follow the layout of the
+existing SPL token accounts to ensure compatibility in RPC when fetching pre and
+post token balances.
+
+For a holding, the first 165 bytes must contain the same information
+as a normal SPL token holding. The following byte must be the type of
+account (ie. `Holding`), and after that, any data is allowed as required by the
+new token program.
+
+The same applies for mints, but only for the first 82 bytes.
+
+The `amount` field in an SPL token holding should contain a balance in order
+to properly integrate with RPC's `preTokenBalances` and `postTokenBalances`.
+
+### Token Program Registry
+
+We need the [token-list](https://github.com/solana-labs/token-list) to include
+vetted SPL token-conforming programs.
+
+TODO consider adding Rust and C versions of the registry for on-chain programs and RPC
+
+TODO consider detailing the process of including a new program
+
+### Runtime
+
+#### Ephemeral Accounts
+
+In order to reduce friction when using read-only instructions on other programs,
+we need the ability for on-chain programs to dynamically create "ephemeral
+accounts", which only contain data, can be passed through CPI to other programs,
+and disappear after the instruction.
+
+For a program using i-tokens, to get the balance of a holding, the program creates
+an ephemeral account with 8 bytes of data, passes it to the i-token program's
+`GetHoldingBalance` instruction, then reads the information back as a `u64`. At the
+end of the instruction, the account disappears.
+
+TODO What's the best approach to implement this?  For example, does an ephemeral
+account even need a public key?
+
+This solution avoids the need to pass additional scratch accounts to almost all
+programs that use tokens.
+
+TODO what limits do we consider for these? A certain number of accounts?
+A total amount of bytes? Regarding security, this breaks the previous
+check that a CPI must use a subset of accounts provided to the calling program.
```

Comment:
my another wildest idea is separation of account data views for internal and external at the invoke context boundary:

_account.data translation between on-chain data and {in-transit/in-tx} data_

In other words, we want to _redefine_ the current spl-token account.data layout is the ABI itself.

1. If i-token holding account is referenced from non-owned program, it first converted from share's amount to token's amount by runtime by triggering specialized instruction in the i-token program (like specifying Trans1ate11111) and passes the translated account.data to the invoking program. This is on-chain => in-transit translation.

2. If the invoking program internally invokes i-token instruction with the i-token account, it's converted back to share's amount. This is in-transit => on-chain translation. The i-token program only deals with share's amount, internally.

3. Finally, after exiting from the invoked i-token program, the invoking program sees another translated view of mutated token accounts after yet another translation.

pro: external programs work as before without `GetHoldingBalance` and its friends. This means there will be single api for both bare tokens and i-tokens. Otherwise, generic token handling program code always must handle the 2 similar api for the two.
con: might be confusing. more tx processing load due to extra translations?


my observation is that tokens are special beast in the blockchain ecosystem: We want it to behave similarly across its breeds, while we want to customize it as much as possible for the added value at the same time. So, let's take this property to the extreme: minimally acceptable compatible interface is the abi itself of the bare token data.

This is like virtualization and duck-typing a bit.

Reply:
This is a really good idea too!  It would be extremely useful to have a unifying representation that can be used across programs, and doing it through these translations is a slick way of re-using code, eliminating the `Get*` instructions, and eliminating all friction for existing token users.  This was actually one of my first thoughts, but I was worried it would be confusing for app developers (as you point out), and instead opted for the simplest possible `Get*` instructions.

Let's see what other people think about it -- I like it! :smile:

Reply:
What's stored in the "standard" part of the i-token holder account data between transactions in this case?  If not the share amount then wouldn't it be the balance at the time of last instruction executed, which could now be wrong/misleading to the user

Reply:
Is the runtime doing the translation, if so how does it know if it should be doing the translation or not?  Does the runtime have to be aware of which accounts are i-token and which are not?

Reply:
@mvines 

> What's stored in the "standard" part of the i-token holder account data between transactions in this case?

I meant the "standard" part of the i-token holder account will be in token's amount, not share's amount. Overall of my review comments, I wanted to hide the share's amount thing as much as possible, like ERC20. I thought ERC20's approch is right because exposing share's amount outside its own program might too confusing.

>  If not the share amount then wouldn't it be the balance at the time of last instruction executed, which could now be wrong/misleading to the user

Between transactions, I think this is fine. After all, i-tokens's balance is expected to continuously increasing? One concern is to transfer all of it (thus, I proposed `transferCheckedAll`) I think aTokens are seeing similar situation.

Also, as for between instructions and cpi invocations inside a single TX, i-token's token & share amount shouldn't change unless mutated so. In other words, time must not progress inside a tx. I think this is safe assumption, considering atomicity and ACID guarantee.

Reply:
> Is the runtime doing the translation, if so how does it know if it should be doing the translation or not? Does the runtime have to be aware of which accounts are i-token and which are not?

@jackcmay that's good questions. I think there are several options to indicate the flag like this:

- Make runtime recognize wrapper Loader address TranslatedBPFLoader111. (bad orthogonality though, for upgradable/non-upgradable)
- Sneak some flag into the elf header?
- or some kind of marker symbol in the elf symbol table (or even the entrypoint of the translation routine)?

Reply:
> After all, i-tokens's balance is expected to continuously increasing?

It could decrease :)
Imagine an iToken that self-destructs after N epochs, or loses value on each transfer.  An NTF iToken might be designed to work this way.   Your share remains the same but the balance decreases.